openjdk-benchmarks/src/jmh/java/org/conscrypt/benchmarks/SslEngineBenchmark.java

```diff
@@ -0,0 +1,340 @@
+
+            private SSLContext newContext() {
+                try {
+                    return SSLContext.getInstance(PROTOCOL_TLS_V1_2, new OpenSSLProvider());
+                } catch (NoSuchAlgorithmException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        },
+        NETTY {
+            private final SslContext clientContext = newClientContext();
+            private final SslContext serverContext = newServerContext();
+
+            @Override
+            SSLEngine newClientEngine(String cipher) {
+                return initEngine(
+                        clientContext.newEngine(UnpooledByteBufAllocator.DEFAULT), cipher, true);
+            }
+
+            @Override
+            SSLEngine newServerEngine(String cipher) {
+                return initEngine(
+                        serverContext.newEngine(UnpooledByteBufAllocator.DEFAULT), cipher, false);
+            }
+
+            private SslContext newClientContext() {
+                try {
+                    File clientCert = Util.loadCert("ca.pem");
+                    return SslContextBuilder.forClient()
+                            .sslProvider(io.netty.handler.ssl.SslProvider.OPENSSL)
+                            .trustManager(clientCert)
+                            .build();
+                } catch (SSLException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+
+            private SslContext newServerContext() {
+                try {
+                    File serverCert = Util.loadCert("server1.pem");
+                    File serverKey = Util.loadCert("server1.key");
+                    return SslContextBuilder.forServer(serverCert, serverKey)
+                            .sslProvider(io.netty.handler.ssl.SslProvider.OPENSSL)
+                            .build();
+                } catch (SSLException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        };
+
+        abstract SSLEngine newClientEngine(String cipher);
+        abstract SSLEngine newServerEngine(String cipher);
+
+        final SSLContext initContext(SSLContext context, boolean client) {
+            if (client) {
+                File cert = Util.loadCert("ca.pem");
+                context = Util.initClientSslContext(context, cert);
+            } else {
+                File cert = Util.loadCert("server1.pem");
+                File key = Util.loadCert("server1.key");
+                context = Util.initServerContext(context, cert, key);
+            }
+            return context;
+        }
+
+        final SSLEngine initEngine(SSLEngine engine, String cipher, boolean client) {
+            engine.setEnabledProtocols(new String[] {PROTOCOL_TLS_V1_2});
+            engine.setEnabledCipherSuites(new String[] {cipher});
+            engine.setUseClientMode(client);
+            return engine;
+        }
+    }
+
+    public enum BufferType {
+        HEAP {
```

Comment:
@kruton unfortunately, you can't extend `ByteBuffer` in openjdk. You have to define a subclass in the `java.nio` package because you have to implement package-private methods. However, you can't define new classes in `java.nio` because the classloader disallows it:
```
java.lang.SecurityException: Prohibited package name: java.nio
```

Looks like there is no way to do this: http://stackoverflow.com/questions/624458/extending-bytebuffer-class

We'll have to live without that check. I believe what you're trying to do really only applies to Android, so perhaps we benchmark this separately when we address #10.